### ISSUE TITLE
Memory Hungry

### ISSUE THREAD

**Issue by ngoel17** [lhotse-speech/lhotse#556]
scripts such as compute_fbank_musan.py consume 3.6GB of memory. 
This limits how many jobs can be run in parallel, despite the number of available CPUs. 
On another note - lhotse kaldi import $data_dir/train 8000 data/manifests/train
kind of commands also run a little slow.

**Comment by pzelasko** [comment]
It's possible to leverage lazy cuts in Lhotse to reduce the memory overhead. You can use the following:
- creating cuts from recordings and supervisions: [`lhotse.cut.create_cut_set_lazy`](https://github.com/lhotse-speech/lhotse/blob/f1b66b8a8db2ea93e87dcb9db3991f6dd473b89d/lhotse/cut.py#L5063) (make sure to read the docs as it assumes specific sorting)
- extracting features with lazy cuts and iterative writing: [`CutSet.compute_and_store_features_batch`](https://github.com/lhotse-speech/lhotse/blob/master/lhotse/cut.py#L4101)

As to Kaldi data dir imports, I don't think there's too much we could do -- it depends on the data dir size.

**Comment by danpovey** [comment]
Nagenrda told me on a call that he was expecting an import command from a Kaldi data dir to take about a minute and it took something like half an hour.  Piotr, what would likely be the limiting factor in such a command?  Does it access the individual data files somehow, e.g. to validate that they exist or to check their length?  (I imagine getting the num-samples in a file might potentially take some time, especially if the wav.scp was based on a pipe).

**Comment by pzelasko** [comment]
That’d be the most likely explanation. There’s a num jobs argument that can help speed this up https://github.com/lhotse-speech/lhotse/blob/f1b66b8a8db2ea93e87dcb9db3991f6dd473b89d/lhotse/kaldi.py#L60

**Comment by pzelasko** [comment]
> BTW - I don't think it would help to increase the jobs for that scenario, because the speed will be disk-io limited.

If Kaldi's `utils/data/get_reco2dur.sh` or `steps/make_mfcc.sh` gets faster with more jobs then this would be as well; I've seen solid speedups even on fairly slow grids.

> ok. If it's verifying then that would make sense. But it does demand existence of "segments" files for creating supervision.

That's correct.

**Comment by armusc** [comment]
> > BTW - I don't think it would help to increase the jobs for that scenario, because the speed will be disk-io limited.
> 
> If Kaldi's `utils/data/get_reco2dur.sh` or `steps/make_mfcc.sh` gets faster with more jobs then this would be as well; I've seen solid speedups even on fairly slow grids.
> 
> > ok. If it's verifying then that would make sense. But it does demand existence of "segments" files for creating supervision.
> 
> That's correct.

what it's taking much more time w.r.t. to Kaldi is then the feature extraction after the import; I guess it's re-opening the same file each time the feature is computed for each segment within the file (in my case, I had to use trim_to_supervisions after the import, so I have a "cut per segment")
but is there a reason why the channel is not supported? the Kaldi segment file with channel indication (i.e. 0/1 in the last field) is not accepted, at the moment I have to use only mono-channel dataset, I know you have been working on multi-channel support, but it would have been fine for a few tests with just a way to treat the second channel segments with their own cut

**Comment by danpovey** [comment]
I think the reason is that channel in segments file is not supported in Kaldi's data dir.
But we could support it in Lhotse as an extension, maybe.

**Comment by armusc** [comment]
> I think the reason is that channel in segments file is not supported in Kaldi's data dir. But we could support it in Lhotse as an extension, maybe.

really? I thought there's always been a fifth field with 0/1 in the segment file and extract segment would use that info to identify the channel when computing features; so I guess I did it myself and I don't even remember...

**Comment by danpovey** [comment]
That has always been supported by extract-segments, but not by the data-dir format.  It would fail validation.
There are reasons for this: it's supposed to be differentiated in the "recording".

**Comment by pzelasko** [comment]
> what it's taking much more time w.r.t. to Kaldi is then the feature extraction after the import; I guess it's re-opening the same file each time the feature is computed for each segment within the file (in my case, I had to use trim_to_supervisions after the import, so I have a "cut per segment")

You can run feature extraction first and then call `trim_to_supervisions`, the cuts will correctly read just the relevant subset of features extracted for the full recording. Make sure you are using either `LilcomChunkyWriter` (default) or `ChunkedLilcomHdf5Writer` for saving features, because other writers might case the reads to be quite inefficient.

**Comment by csukuangfj** [comment]
> As to Kaldi data dir imports, I don't think there's too much we could do -- it depends on the data dir size.

I am not sure if you are willing to replace the Python implementation [kaldiio](https://github.com/nttcslab-sp/kaldiio)
with a C++ implementation [kaldi_native_io](https://github.com/csukuangfj/kaldi_native_io) and to test the speed difference for large dataset.

**Comment by pzelasko** [comment]
Interesting project! I’m open to replacing it but might be tough for me to find the time right now. If you guys need it please make a PR.

**Comment by pzelasko** [comment]
Is this issue resolved, or is there anything else we can do?

**Comment by ngoel17** [comment]
Should I check the PR that @Fangjun suggested? Sorry I didn't keep track. I
am importing 5 k hours today from kaldi to lhotse.
It has been running for a couple of hours now. How should I test if it's
resolved?

**Comment by pzelasko** [comment]
You can pull latest master, install Fangjun’s native kaldiio library, set a higher number of jobs for the import script, and let us know if it’s better then.

**Comment by csukuangfj** [comment]
If it turns out importing data from Kaldi still takes a lot of time after the fix, there is still something we can do.
https://github.com/lhotse-speech/lhotse/blob/6dd9d6aacb8eed8461f795049c31b79153f6dbde/lhotse/kaldi.py#L144-L145

As the code only requires the metadata of a matrix, there is no need to allocate memory and copy data to decompress and read the matrix, we can return the shape information of the matrix by reading only the header.

**Comment by danpovey** [comment]
It should be possible to update the kaldi_native_io project to be able to read just the metadata, e.g. by adding an option to skip the actual data-reading (at least in cases where the data is binary).  This may be a good TODO for someone less experienced, if any such people follow this project.

**Comment by ngoel17** [comment]
I actually ran into a different problem. The data sources are mixed, and
one of those is swbd in kaldi format. I just want to use my version instead
of asking lhotse to prepare swbd. So the wav.sch looks like this....
   sw02001-A /home/ngoel/kaldi/tools/sph2pipe_v2.5/sph2pipe -f wav -p -c 1
/mnt/disk1/LDC97S62/swb1_d1/data/sw02001.sph |

  and I get the error message(s) - Failed while writing sample data to
(null)

  I also think that it should be allowed for some "text" entries to be
empty (i.e. pure silence in context of speech).

**Comment by ngoel17** [comment]
also the command  
lhotse prepare fisher-english --absolute-paths True -j 20 . ~/work/data/manifests/fisher-english
gives the error
NotADirectoryError: [Errno 20] Not a directory: 'LDC2004S13/LDC2004S13/fisher_eng_tr_sp_LDC2004S13.zip.001/audio'
and the structure at my end is:
ls LDC2004S13/LDC2004S13
fisher_eng_tr_sp_LDC2004S13.zip.001  fisher_eng_tr_sp_d1  fisher_eng_tr_sp_d3  fisher_eng_tr_sp_d5  fisher_eng_tr_sp_d7
fisher_eng_tr_sp_LDC2004S13.zip.002  fisher_eng_tr_sp_d2  fisher_eng_tr_sp_d4  fisher_eng_tr_sp_d6

**Comment by csukuangfj** [comment]
>   and I get the error message(s) - Failed while writing sample data to
(null)

Are there more error logs, e.g., the stacktrace?

**Comment by ngoel17** [comment]
No the code does not crash.

**Comment by csukuangfj** [comment]
That warning is from
![Screen Shot 2022-02-24 at 08 24 27](https://user-images.githubusercontent.com/5284924/155433521-ff5bea96-d39a-4218-8aea-2a3ba39096ee.png)

There is a similar question in Kaldi's google help group
https://groups.google.com/g/kaldi-help/c/9qW7Z2yP6Sw

![Screen Shot 2022-02-24 at 08 26 03](https://user-images.githubusercontent.com/5284924/155433627-72463303-28eb-4cce-97cf-74773494daea.png)


The reply from Dan is
![Screen Shot 2022-02-24 at 08 26 39](https://user-images.githubusercontent.com/5284924/155433673-2c1505b7-de91-45cb-9c61-52a167f5c157.png)

**Comment by csukuangfj** [comment]
> NotADirectoryError: [Errno 20] Not a directory: 'LDC2004S13/LDC2004S13/fisher_eng_tr_sp_LDC2004S13.zip.001/audio'

That error is caused by
https://github.com/lhotse-speech/lhotse/blob/88d8d964d8323b20693304bc1c3e9d88a204bd26/lhotse/recipes/fisher_english.py#L154-L158

You can check whether your dataset directory layout is compatible with the above code.

**Comment by ngoel17** [comment]
Thanks!! It worked. I see that there is a significant speedup as I set nj=20 on my 8 core machine, 
and I see about 300% CPU utilization.  (I don't have pre-computed features BTW). 
 towards the end, the python script spends a significant amount of time at 100% utilization while its writing manifest. 
As I extract features (icefall - compute_fbank.py) it first loads stuff and the memory utilization grows to about 6 GB pretty 
rapidly, and then it continues to increase more slowly to 15 GB and beyond. My current machine limit is 30GB. 
My recordings manifests is 14MB and supervisions manifest is 128MB. 
I finally get this message. Looking at the CPU usage pattern, I believe that feature extraction did not start at this time.

Extracting and storing features (chunks progress):   0%|                                                                                                    | 0/8 [00:54<?, ?it/s]
Traceback (most recent call last):
  File "/mnt/dsk2/icefall/egs/dk2/./local/compute_fbank.py", line 98, in <module>
    compute_fbank()
  File "/mnt/dsk2/icefall/egs/dk2/./local/compute_fbank.py", line 80, in compute_fbank
    cut_set = cut_set.compute_and_store_features(
  File "/home/ngoel/lhotse/lhotse/cut.py", line 4456, in compute_and_store_features
    cuts_with_feats = combine(progress(f.result() for f in futures))
  File "/home/ngoel/lhotse/lhotse/manipulation.py", line 30, in combine
    return reduce(add, manifests)
  File "/home/ngoel/.local/lib/python3.9/site-packages/tqdm/std.py", line 1180, in __iter__
    for obj in iterable:
  File "/home/ngoel/lhotse/lhotse/cut.py", line 4456, in <genexpr>
    cuts_with_feats = combine(progress(f.result() for f in futures))
  File "/home/ngoel/anaconda3/envs/k2/lib/python3.9/concurrent/futures/_base.py", line 445, in result
    return self.__get_result()
  File "/home/ngoel/anaconda3/envs/k2/lib/python3.9/concurrent/futures/_base.py", line 390, in __get_result
    raise self._exception
concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.

**Comment by ngoel17** [comment]
even if I change 80 to 1 in this line 
   https://github.com/k2-fsa/icefall/blob/ac7c2d84bc44dd83a5653f68677ae1ef16551eea/egs/librispeech/ASR/local/compute_fbank_librispeech.py#L86
 in the equivalent code (just changing the manifests otherwise), the code breaks with the same issue.

**Comment by ngoel17** [comment]
Even if I set num_jobs=1 in the equivalent of 
https://github.com/k2-fsa/icefall/blob/ac7c2d84bc44dd83a5653f68677ae1ef16551eea/egs/librispeech/ASR/local/compute_fbank_librispeech.py#L47
the memory consumption increases to 18.4 GB before feature extraction begins.

**Comment by csukuangfj** [comment]
> My recordings manifests is 14MB and supervisions manifest is 128MB.

I guess your training data is probably more than 4k hours.  You may need to split your manifests into smaller pieces before pre-computing features or use on-the-fly feature extraction during training.

**Comment by csukuangfj** [comment]
There are discussions about the CPU OOM issue for large datasets (e.g., 10k hours) in https://github.com/k2-fsa/icefall/pull/120

**Comment by pzelasko** [comment]
Icefall's LibriSpeech recipe uses Lhotse in a way that is suitable for small and medium sized datasets -- it reads the whole manifest into memory and does various operations on it. But this approach doesn't scale to larger datasets. We have multiple utilities for working with large data: sequential manifest writers (`CutSet.open_writer()`), lazy manifest readers (`CutSet.from_jsonl_lazy`), dynamic samplers (`DynamicCutSampler` and `DynamicBucketingSampler`), conversion to sequential I/O (`export_to_webdataset`/`CutSet.from_webdataset`), and probably others. Each of those has documentation (typically with examples) on its own, but we're currently some sort of top-level doc/tutorial that would put these pieces into one place for easier discoverability. 

If you're working with large datasets, you'll definitely want to explore using the tools mentioned above.

**Comment by pzelasko** [comment]
Oh and spare yourself the trouble and avoid using Hdf5 for storing features -- if you see a line like `storage_type=LilcomHdf5Writer` or `storage_type=ChunkedLilcomHdf5Writer`, just remove it, and it'll use the default storage type for features which doesn't have memory leak issues.

**Comment by ngoel17** [comment]
Will do on removing HDF5. Does the inbuilt fisher prep recipe treat fisher
as a large dataset? Maybe that can be used as a prototype? I would not know
what advantage anybody gets from treating small data as small. Shouldn't
all data be treated as large data?

**Comment by pzelasko** [comment]
Yeah, ideally all recipes would be using the lazy cuts stuff, but that was only added at some point later once the recipes already existed. Would be great if somebody had the time to update them. 

There is a running but unfinished PR with SWBD + Fisher recipe I pushed in Icefall, I only trained a "pure SWBD" model, but it shows how to run a training with on-the-fly feature extraction where the data is treated as "large data" and has a limited memory usage at most (all?) times. If you find it interesting maybe you'd like to finish it - it needs WER evaluation vs eval2000 and for somebody to run Fisher+SWBD training, and possibly some hparam tweaking (it uses Libri setup) https://github.com/k2-fsa/icefall/pull/184/files

**Comment by armusc** [comment]
Hi 

I have a corpus of about 5k of effective speech
Cuts count: 5933260
Total duration (hours): 5318.5
Speech duration (hours): 5318.5 (100.0%)
***
Duration statistics (seconds):
mean    3.2
std     4.0
min     0.1
25%     1.0
50%     1.9
75%     3.9
max     246.6

which is imported from kaldi, features included (indeed it's one corpus augmented 8 times)
because of CPU OOM issues, I followed the recommendations in this thread
1) filtered segments 1 < duration < 20
2) used load_manifest_lazy
3) used DynamicBucketingSampler(CutSampler)

I don't know if there's something else I forgot.
But nothing changes with respect to the set up of the standard librispeech, the memory is all consumed
I can see that training blocks immediately in

train_one_epoch, at the line
for batch_idx, batch in enumerate(train_dl):     

I left a log statement within the __next__ method of the
class LazyJsonlIterator:

and I see that 
1529350 of the 5933260 cuts are in the batch

after that, memory grows gradually until all memory (256GB) is consumed and the program has not advanced further within the batch loop in train_one_epoch

did that look a strange behavior to you? suggestions where I should look into to understand where the issue lies?

thanks in advance

**Comment by armusc** [comment]
when
train_sampler = DynamicBucketingSampler(
is called in asr_datamodule.py
I can see that 11000 cuts are selected (because of a log left in __next__ )

it's the first time
for batch_idx, batch in enumerate(train_dl):
is called that I see the log being printed ~1.5 M times
don't know if this makes sense
trying to understand what happens

**Comment by pzelasko** [comment]
Try replacing it with DynamicCutSampler and see if the issue persists — as a sanity check. Also try experimenting with a larger buffer_size arg, maybe the buckets are under filled because of short duration cuts in your data?

**Comment by armusc** [comment]
Nothing has changed since I started experimenting with this, i.e. using lazy loading and DynamicBucketing or DynamicSampler
the buffer_size params in DynamicCutSampler is  shuffle_buffer_size? it's the only param I see with buffer_size in the name
btw, it is written that increasing the size might further increase memory usage

SHould not "max_cuts" be used to limit the number of cuts in the batch? anyway, setting to 10000 has not changed anything, according to enumerate(train_dl) there are 1.5M cuts in the batch with very negligible changes in this number as I modify some of these params
I tried, for the sake of testing, to use the mach smaller validation corpus for training and I can adavance in the training (even though the memory occupation is still high, i.e. 18% of server memory for each of 4 CPU processes, resulting from num_workers=2 and world_size=2, which means that 70% of the 256GB memory is constantly taken)
at the moment,  I can not do much

according to this setting, the number of cuts in enumerate(train_dl) in train_one_epoch should be much lower, right?

**Comment by pzelasko** [comment]
Thanks for a detailed description. I’m on vacation right now and will be able to help you more one week from now.

**Comment by ngoel17** [comment]
@pzelasko I understand that you would like us to use CutSet.from_json_lazy() and DynamicCutSampler(). However, this issue of memory still perplexes me.  In the case of fisher, for example, the supervisions and recordings manifest take about 80GB disk Compressed (gz) and 500MB uncompressed. But when loaded in memory, it's 4.5GB. If the manifest could be more efficient in memory that would help even in the case of dynamic cuts and when we are dealing with complex shuffling scenarios.

**Comment by danpovey** [comment]
I wonder whether the individual cuts could perhaps be stored as a string, and only turned to an object on demand...

**Comment by pzelasko** [comment]
I’ll think about it but I’m not sure how we could really optimize the memory usage further at this point.. 

> I wonder whether the individual cuts could perhaps be stored as a string, and only turned to an object on demand...

I can’t think of a way to code that without breaking something else right now. But maybe it can be done.

**Comment by ngoel17** [comment]
Okay, I understand the issue a little better now after reading this reference. https://pythonspeed.com/articles/python-integers-memory.

**Comment by armusc** [comment]
I would be happy to see already difference in the use of lazy_load and Dynamic Bucketing; which I don't see.
Could you confirm that everything related  to reading manifest and loading data is in:
serialization.py
dataset/sampling/
I am using pytorch 1.8

for example, in my dataset of 320h, I am setting 3 workers, 2 GPU, other parameters for asr_datamodule
num buckets = 30
shuffle = True
max_duration = 100 
buffer_size = 10000
(I also tried smaller values for those)

the features for this corpus amount to 18GB
top shows 6 CPU processes, each one taking 12% of the server RAM, that is about 30GB, for a total of 70% of CPU server consumed constantly, since the start of the training

this happens whenever I use:
train_sampler = DynamicBucketingSampler(    and    cuts_train = CutSet.from_jsonl_lazy(

and  

train_sampler = SingleCutSampler(    and      cuts_train = CutSet.from_file(
regardless

I cannot use bigger dataset, in the order of a few thousandths of hours even with one just one worker and one GPU because all the RAM would be filled up very soon
but I guess I'm doing something wrong here, because you are all seeing a different behavior by using lazy loading and Dynamic Bucketing
any tip where I should look at?

**Comment by pzelasko** [comment]
Ok let’s take it step by step. 

1. Verify that the extension of the file is .jsonl or .jsonl.gz and it has one record per line. 

2. Open interpreter (python/ipython) and open the CutSet with “from_jsonl_lazy”. Open a new terminal and run htop, find the interpreter process, report the memory usage at this point. 

3. In the interpreter, run a simple iteration: “for cut in cuts: pass”. Keep looking at the memory usage in htop (should be constant).

4. Create a DynamicBucketingSampler, observe the memory usage in htop (should go up a bit)

5. Iterate DynamicBucketingSampler: “for batch in sampler: pass”. Monitor the memory usage in htop. 

Please tell me at which step it starts to be excessive (if any).

**Comment by armusc** [comment]
thanks for your help

1) yes, the input manifest has extension jsonl.gz; it contains one record per line, just to be clear each line represents a MonoCut, which represents a segment of multi-channel multi-segment audio recording; I did a trim_to_supervisions of the CutSet read from the input manifest

from 2)  to 5) the memory usage is always the same; it's  ~1GB, that is 0.4% of the server memory for the only python process created when the manifest was read, i.e.: cuts_train = CutSet.from_jsonl_lazy(....)
it has never went up over that

btw, it's probably not relevant to the discussion, but this is the difference I see when using lazy loading and Dynamic Bucketing
![tensorboard_engcts_kaldifeat_normalized_lazyeiger](https://user-images.githubusercontent.com/46787089/157669431-86fad608-7125-4f53-ad78-7f1d5bb48bac.png)

you can see this alternance of peaks and valleys in the losses, that valley happens at the begninning of an epoch, this is using lzy loading and DynamicBucketing, then you see that I re-started the training at epoch 16 with no lazy loading and SImpleCutSampler and there's no alternance of peaks and valleys (this is the only thing I am seeing)

**Comment by danpovey** [comment]
The peaks-and-valleys thing is an issue that I thought we had solved, but I don't fully recall the details; it is something to do with buckets of certain sizes being exhausted earlier than others.  IIRC the number of buckets is relevant.

**Comment by pzelasko** [comment]
@armusc in that case, the lazy loading mechanism seems to work correctly. Can you wrap it into a script that adds K2 dataset and a dataloader and just iterate over the dataloader? And see if the memory is OK or starts growing. 

Also — did you precompute the features and store them in HDF5? It tends to blow up the memory after some time, but you can fix it by either using the new default storage type (LilcomChunkyWriter) or on the fly features.

**Comment by armusc** [comment]
sorry for the late reply, I'm experiencing connection issues with the servers

I did what you suggested; beside the one process that stays at 0.4% server memory, there is another one that works at 100% that stays constantlt at 1.4% server memory (about 3.5GB)
so iterating over the dataloader is not the problem
as for the feature, in this specific case it's kaldi imported feature
otherwise, yes, while doing feature extraction in lhotse I stored with hdf5

**Comment by armusc** [comment]
no, sorry, I take back what I just wrote
that dataset was the same pone but not augmented; it's 40h of speech

I relaunched with that same dataset augmented 8 times which brings the overall size to about 320h and indeed the memory occupation it's about 30GB, which is 11.4% of the server memory; this is by iterating over the data loader, non further processing

**Comment by pzelasko** [comment]
BTW I skimmed through `kaldi_native_io` code, it seems that it is reading the whole scp file into memory. I don't know if that could be related to the issue? You can try to run the iteration with the branch in PR https://github.com/lhotse-speech/lhotse/pull/612

But I expect if you either switch to on-the-fly features or LilcomChunkyWriter storage type, the issue will magically disappear. You can use `lhotse copy-feats` to change the feature storage from Kaldi scp+ark to `lilcom_chunky`.

**Comment by armusc** [comment]
> BTW I skimmed through `kaldi_native_io` code, it seems that it is reading the whole scp file into memory. I don't know if that could be related to the issue? You can try to run the iteration with the branch in PR #612
> 
> But I expect if you either switch to on-the-fly features or LilcomChunkyWriter storage type, the issue will magically disappear. You can use `lhotse copy-feats` to change the feature storage from Kaldi scp+ark to `lilcom_chunky`.

1) Copying the features from Kaldi to LilcomChunkyWriter is doing the job, indeed. I tried with different subsets of the training corpus and the full corpus and the memory occupation is the same, about two processes that take 1.6% of server RAM and 4 additional ones that take less. For comparison, with the same features before, I had 12% of CPU memory for each of 6 processes. MOst important, RAM consumption is the same regardless of the size of corpus.
ANother very good news is that training is much faster, like 10 times faster; the loss curves would be the same as posted in this thread (that one with the alternance of peaks and valleys) same batches, but now each epoch is less than 20 minutes, before it was 3 hours; I can see the GPUs are always working, before they were at 0% most of the time

2) I tried the features on-the-fly, but this is not an option that is suitable to me; it is slow, and features must be re-computed at each epoch and for different trainings; 

3) I tried simply computing the features with lhotse, but I have OOM problems during the computation; it's a standard Fbank feature extractor; dispatching the computation over 32 CPUs, memory consumption slowly increases for each process until all memory is taken (256GB). Am I the only one to have observed this? as far as I know, computation is done on python and then with torchaudio, there can't be a memory leak, I guess

anyway, approach 1) is working fine for me.
thanks for the help

**Comment by pzelasko** [comment]
I'm glad we were able to resolve it, thanks for your patience!!

As for 1, it's interesting to know that Kaldi feature readers are so memory hungry and slow in this context. I'm not sure if there's anything we can do about it TBH... We might want to update the Kaldi import with an option to copy the features into another support format at least.

As for 2 -- that's right, unless you have very fast disks, on-the-fly features may slow you down a lot.

As for 3, I think we are dealing with manifests sub-optimally in Lhotse's `compute_and_store_features` -- it keeps some things in memory that it technically doesn't have to. I might update it later, just want to make sure I won't break existing workflows that work with smaller data without issues.

There is actually an option 4 if you have extra storage space and I/O is a big issue, for which I've just finished writing a tutorial here: https://github.com/lhotse-speech/lhotse/blob/master/examples/02-webdataset-integration.ipynb

**Comment by armusc** [comment]
if I can add, since I have seen that several methods (split, filter etc) have been made lazy, this become also necessary for the copy-feats 
even if I split the manifest for a big dataset into smaller manifests, then the original storage feature file (i.e. feats.scp for kaldi) is charged entirely for each splitted manifest; if I had already a kaldi feature file too big to stay in memory, then there's no chance to use copy-feats successfully (even with only one split)

**Comment by pzelasko** [comment]
I think in that case you'd best split Kaldi data dir into several smaller dirs, import separately, copy feats separately, and merge into one manifest later (or lazily add them). The code that reads feats.scp into memory is outside of Lhotse and I don't think we can do anything to amend that on Lhotse side.

**Comment by armusc** [comment]
right, that's true

**Comment by armusc** [comment]
Hi

I tried what has been suggested; at the end, when I have to combine the partial manifest into a single one, I get
    merged_cuts = {**self.cuts, **other.cuts}
TypeError: 'LazyJsonlIterator' object is not subscriptable

when using jsonl.gz files

if using json files the combination works
is this the expected behaviour?

**Comment by armusc** [comment]
my bad, updated cut.py; no more error now

**Comment by RuABraun** [comment]
@pzelasko  Sorry to resurrect an old topic, but I've encountered this issue now (very high mem usage) and am wondering what's causing it (was using `LilcomChunkyWriter` and `BucketingSampler`, trying out `DynamicBucketingSampler` now). I've skimmed over the previous comments on this and one thing that's not clear to me is why a CutSet should be causing large memory usage since they don't actually take much memory? 

More specifically the lazy manifest feature seems to be to avoid loading an entire CutSet into memory, same with using a DynamicBucketingSampler over BucketingSampler. But in my case 1.8M cuts are 280M. That's not a lot. Of course if you were to load and keep in memory the features/audio-arrays associated with the cuts that would be a lot, but that is not done AFAIK (they're only loaded when __getitem__ of the Dataset is called and they're not saved). 

So what could be causing the large memory usage and how does DynamicBucketingSampler help with that since a CutSet does not actually use a lot of memory? 

By the way, iterating across the training BucketingSampler does not cause a significant increase in memory usage, for some reason that happens at the end when computing the validation loss (the set is tiny) (goes from 15G to what seems to be 120G).

**Comment by danpovey** [comment]
It is possible that you are keeping the validation features in memory somehow?  Is the dataloader for validation the same type as for train?  And are the underlying features on disk the same type?  (there are memory leaks in the hdf5 reader)

**Comment by danpovey** [comment]
If you mean the problem disappears when you decrease the validation interval significantly, that is very strange.
When you look at the memory usage, make sure you look at resident memory.  The reported virtual memory can be huge if using GPUs, but that is not meaningful.  If you can find a way to plot the memory usage in a fine grained way, or report it from python, it would be good to see whether it linearly increases while computing the validation loss.
Also, and this is a long shot, I can imagine that if you were not using a GPU and you had done something funky like setting requires_grad = True on something and the loss was being accumulated as a tensor, it might be keeping a lot of the graph around in memory and the tensors would use a lot of memory.

**Comment by pzelasko** [comment]
Yeah seeing some sort of memory use plot would be great. There is a simple [memory profiler](https://pypi.org/project/memory-profiler/) that with a bit of code annotations can provide this plot with markers when certain things happen (e.g. start of validation, end of validation). 

Regarding CutSet memory usage: lazy CutSet is super lean, it's just a path to file/other resource. Eager CutSet loads all manifests into memory and for large datasets the overhead of creating all these Python objects is annoying (long loading time, possible memory hungry). BucketingSampler pre-computes the buckets and by design needs to hold all of those cuts in memory. It's more suitable for small to medium data. DynamicBucketingSampler only reads a sample of Cuts to compute buckets and doesn't hold them in memory; during iteration it only keeps a `buffer_size` number of cuts in the memory and as it yields batches, it refills the buffer and puts the cuts into appropriate buckets. I hope that clears up why some things are more or less memory efficient.
